File: cases/plugin/ladcraft-r7_new/src/apply/cell-format.ts
/**
 * Cell open-sheet format ops applied via Asc.plugin.callCommand + viewport nudge.
 * @see cases/knowledge-base/r7-api-handoff/r7-cell-format-selection-font.md
 */

export type CellFormatTarget = "used" | "selection" | string;

export interface CellFormatSpec {
  bold?: boolean;
  italic?: boolean;
  fontName?: string;
  fontSize?: number;
  /** RGB 0..255 */
  fontColorRgb?: [number, number, number];
  fillRgb?: [number, number, number];
  underline?: string;
  /** First row of used range (header styling). */
  headerBold?: boolean;
  headerFillRgb?: [number, number, number];
  bordersOutline?: boolean;
}

export interface CellFormatProposal {
  target?: CellFormatTarget;
  format: CellFormatSpec;
}

/** Infer a practical format from a short user phrase (no proposal). */
export function inferCellFormatFromUserText(userText: string): CellFormatProposal {
  const t = String(userText || "").toLowerCase();
  const format: CellFormatSpec = {};
  if (/шапк|header|заголов/.test(t)) {
    format.headerBold = true;
    format.headerFillRgb = [240, 240, 240];
  }
  if (/жирн/.test(t) && !format.headerBold) {
    format.bold = true;
  }
  if (/жирн/.test(t) && /шапк|header|заголов/.test(t)) {
    format.headerBold = true;
  }
  if (/подчеркн/.test(t)) {
    format.underline = "single";
  }
  if (/arial/i.test(userText)) {
    format.fontName = "Arial";
  }
  if (/заливка|серым|серую/.test(t) && !format.headerFillRgb) {
    format.fillRgb = [240, 240, 240];
  }
  if (!Object.keys(format).length) {
    format.headerBold = true;
    format.headerFillRgb = [240, 240, 240];
    format.fontName = "Arial";
    format.fontSize = 11;
  }
  let target: CellFormatTarget = "used";
  if (/выделен|selection|выдел/.test(t)) target = "selection";
  return { target, format };
}

File: cases/approved-variants/excel-pivot-ok-14-53-24-07/skills/excel_pivot_toolkit/scripts/cell_format.py
async def handler(state, params):
    """Prepare open-Cell format op; plugin applies via Asc."""
    params = params if isinstance(params, dict) else {}
    target = get_string(params, "target").strip().lower() or "used"
    if target not in ("used", "selection", "range", "cells"):
        return {"ok": False, "error": "target: used | selection | range | cells"}
    range_addr = get_string(params, "range").strip()
    cells_raw = get_array(params, "cells")
    cells = [str(x or "").strip() for x in cells_raw if str(x or "").strip()]
    if target == "range" and not range_addr:
        return {"ok": False, "error": "range обязателен при target=range"}
    if target == "cells" and not cells:
        return {"ok": False, "error": "cells обязателен при target=cells"}

    fmt_in = params.get("format")
    if not isinstance(fmt_in, dict) or not fmt_in:
        return {"ok": False, "error": "format обязателен (объект со свойствами оформления)"}

    allowed = {
        "bold",
        "italic",
        "fontName",
        "fontSize",
        "fontColorRgb",
        "fillRgb",
        "underline",
        "headerBold",
        "headerFillRgb",
        "bordersOutline",
    }
    unsupported = []
    format_out = {}
    for key, val in fmt_in.items():
        if key not in allowed:
            unsupported.append(str(key))
            continue
        format_out[key] = val

    # Safety: «шапка жирным» на used → headerBold (строка 1), не bold всего листа/столбца.
    if (
        target == "used"
        and format_out.get("bold") is True
        and "headerBold" not in format_out
        and set(format_out.keys()) <= {"bold"}
    ):
        format_out = {"headerBold": True}
    if format_out.get("headerBold") is True and "bold" in format_out:
        del format_out["bold"]

    if not format_out:
        return {
            "ok": False,
            "error": (
                "format пуст после фильтрации. Поддерживаются только: "
                + ", ".join(sorted(allowed))
                + ". Не поддерживаются (в т.ч. толщина/внутренние границы): "
                + (", ".join(unsupported) if unsupported else "—")
            ),
            "supportedFormatKeys": sorted(allowed),
            "rejectedKeys": unsupported,
        }

    data = {
        "target": target,
        "range": range_addr or None,
        "cells": cells if cells else None,
        "format": format_out,
        "rejectedKeys": unsupported if unsupported else None,
    }
    return {
        "ok": True,
        "type": "cell_format",
        "data": data,
        "partial": bool(unsupported),
        "rejectedKeys": unsupported if unsupported else None,
        "note": (
            "Часть полей отброшена (не поддерживаются): " + ", ".join(unsupported)
            if unsupported
            else None
        ),
    }
